portability-worker/src/main/java/org/dataportabilityproject/worker/WorkerMain.java

```diff
@@ -24,38 +25,62 @@
 import org.dataportabilityproject.api.launcher.Logger;
 import org.dataportabilityproject.spi.transfer.extension.TransferExtension;
+import org.dataportabilityproject.spi.transfer.provider.Exporter;
+import org.dataportabilityproject.spi.transfer.provider.Importer;
 
 /**
- * Main class to bootstrap a portabilty worker that will operate on a single job whose state is held
- * in WorkerJobMetadata.
+ * Main class to bootstrap a portability worker that will operate on a single job whose state is
+ * held in WorkerJobMetadata.
  */
 public class WorkerMain {
   public static void main(String[] args) throws Exception {
     Thread.setDefaultUncaughtExceptionHandler(UncaughtExceptionHandlers.systemExit());
 
     ExtensionContext extensionContext = getExtensionContext();
 
+    Injector injector = Guice.createInjector(new WorkerModule());
+    WorkerImpl worker = injector.getInstance(WorkerImpl.class);
+
+    // Poll job, and determine its data type, import service, and export service
+    worker.pollForJob();
+    WorkerJobMetadata metadata = worker.getWorkerJobMetadata();
+    String transferDataType = metadata.getDataType();
+    String exportService = metadata.getExportService();
+    String importService = metadata.getImportService();
+    Exporter<?, ?> exporter = null;
+    Importer<?, ?> importer = null;
+
+    // TODO: Next version should ideally not load every TransferExtension impl, look into
+    // solutions where we selectively invoke class loader.
     List<TransferExtension> extensions = new ArrayList<>();
     ServiceLoader.load(TransferExtension.class).iterator().forEachRemaining(extensions::add);
+    Preconditions.checkState(
+        !extensions.isEmpty(), "Could not find any implementations of TransferExtension");
 
     for (TransferExtension extension : extensions) {
-      extension.initialize(extensionContext);
-    }
-    for (TransferExtension extension : extensions) {
-      extension.getExporters();
-      extension.getImporters();
-    }
-    for (TransferExtension extension : extensions) {
-      extension.start();
+      if (extension.getServiceId().equals(exportService)) {
+        extension.initialize(extensionContext);
+        Preconditions.checkState(
+            exporter == null,
+            "Can only instantiate exporter from one matching TransferExtension. Previously found :"
+            + exporter.getClass() + ", and now found: " + extension.getClass());
+        exporter = extension.getExporter(transferDataType);
+      } else if (extension.getServiceId().equals(importService)) {
+        extension.initialize(extensionContext);
+        Preconditions.checkState(
+            importer == null,
+            "Can only instantiate importer from one matching TransferExtension. Previously found :"
+                + importer.getClass() + ", and now found: " + extension.getClass());
+        importer = extension.getImporter(transferDataType);
+      }
     }
 
```

Comment:
As a minor suggestion for a later PR I think lines 57-75 could potentially be shortened by streaming the extension collection, filtering, finding the first matching, and getting or throwing on the returned Optional.

This is slightly different than checking if more than one extension matches.

Reply:
Thanks, I was actually thinking the same thing... I left a TODO for now